### ISSUE TITLE
How to train ConvLSTM model with my own dataset?

### ISSUE THREAD

**Issue by houjanet** [tsugumi-sys/SA-ConvLSTM-Pytorch#93]
Hello! Thank for your work! Could you please tell me how to train ConvLSTM model with my own dataset? I would be grateful if you could reply to me!

To load npy files, I added the following code in moving_mnist.py: 
def MNISTDataLoader(path):
    data = np.load(path)
    data_trans = data.transpose(1, 0, 2, 3)
    return data_trans

class MovingMNISTDataset(Dataset):
    def __init__(self, path):
        self.path = path
        self.data = MNISTDataLoader(path)
    def __len__(self):
        return len(self.data[:, 0, 0, 0])
    def __getitem__(self, index):
        self.trainsample_ = self.data[index, ...]
        self.smaple_ = self.trainsample_ / 255.0
        self.smaple = torch.from_numpy(np.expand_dims(self.smaple_, axis=1)).float()
        return self.smaple

And I changed it to read my data in this way.
        # moving_mnist = MovingMNIST(
        #     root="./data", download=True, transform=lambda x: x / 255.0
        # )
        
        moving_mnist = MovingMNISTDataset('/data/my_paper_lite//data_final_cluster/30npy/半導體業_full.npy')

Since the shape of my data is (20, 1255, 8, 10), I modified the parameters in moving_mnist_convlstm.py as follows:
def main():
    ###
    # Common Params
    ###
    artifact_dir = "./半導體業_test"
    input_seq_length = 10
    train_batch_size = 32
    validation_bath_size = 16
    ###
    # Setup Pipeline
    ###
    model_params: Seq2SeqParams = {
        "input_seq_length": input_seq_length,
        "num_layers": 2,
        "num_kernels": 64,
        "return_sequences": False,
        "convlstm_params": {
            "in_channels": 1,
            "out_channels": 1,
            "kernel_size": (3, 3),
            "padding": "same",
            "activation": "relu",
            # "frame_size": (64, 64),
            "frame_size": (8, 10),
            "weights_initializer": WeightsInitializer.He,
        },
    }

![convlstm](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/151805470/7cb514e9-4f3b-4b72-aa7a-279cd5db1cea)
![moving_mnist_1](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/151805470/96897201-b4a8-4b1e-8b78-44960a3cd105)
![moving_mnist_2](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/151805470/5bb138af-cb3e-414a-9819-925edf7f1183)

My questions are:
1. Since my data is a large npy file merged from npy files of different dates and times, I would like to know what each evaluation.png represents in terms of date and time.
![data](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/151805470/33373749-065e-4b4b-83a5-65f429f831dd)

2. If I want to colorize the images and label each frame with the date and time in the file name, do I need to make modifications in the visualize_utils.py section?

def save_pred_vs_label_images(
    image_path: str, label_frames: torch.Tensor, pred_frames: torch.Tensor
):
    """
    frames: (BatchSize=1, Channels, Frames, H, W)
    """
    os.makedirs(os.path.dirname(image_path), exist_ok=True) # 確保指定的圖像路徑存在，如果不存在則會創建
    num_frames = label_frames.size(2) # 計算標籤影像的幀數量，使用索引 2 從形状元组中獲取第三個維度的大小，即影像的幀數
    fig = plt.figure(figsize=(16, 6))
    subfigs = fig.subfigures(nrows=2, ncols=1) # 將圖形分成兩個子圖形，每個子圖形有一行，列數等於影像的幀數量
    for rowidx, subfig in enumerate(subfigs): # 使用 enumerate() 函式迭代子圖形列表，其中 rowidx 是索引，subfig 是子圖形對象
        title = "prediction" if rowidx == 1 else "label" # 如果 rowidx 等於 1，則標題為 "prediction"，否則為 "label"
        subfig.suptitle(title)
        axs = subfig.subplots(1, num_frames) # 創建子圖形的軸，數量等於影像的幀數量
        frame_data = pred_frames if rowidx == 1 else label_frames # 根據 rowidx 的值，它會從 pred_frames 或 label_frames 中選擇相應的幀數據
        frame_data = frame_data.cpu().detach()
        for fidx in range(num_frames):
            ax = axs[fidx]
            with torch.no_grad():
                ax.imshow(frame_data[0, 0, fidx], cmap="gray") # 將影像幀顯示在每個軸上
            ax.set_xlabel(f"frame{fidx}") # 設置 x 軸標籤為 "frame0"、"frame1" 等
            ax.set(xticklabels=[], yticklabels=[], xticks=[], yticks=[]) # 將 x 和 y 軸的刻度標籤和刻度線設置為空

    plt.savefig(image_path)
    plt.close()
![驗證圖](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/151805470/49d70079-73ec-4ede-bae8-4ab3a828cbec)

3. Regarding the parameters, I would like to ask why "num_kernels" is set to 64?

**Comment by houjanet** [comment]
To ensure that this code segment can be executed successfully, I first converted the downloaded mnist_test_seq.npy file into images, and then converted the images back into an npy file, which is the output.npy.

###########################################################################################
Code for converting to images:
import cv2
import os
import numpy as np
data = np.load('./data/MovingMNIST/mnist_test_seq.npy')
data = data.transpose(1, 0, 2, 3)
data.shape
data.ndim

for index in range(len(data[:, 0, 0, 0])):
    count = 0
    for j in data[index, ]:
        img = np.expand_dims(j, -1)
        print(img.shape)
        filename = './movingmnist1/video_%d/frame_%d.jpg'%(index, count)
        count+=1
        savepath = './movingmnist1/video_%d'%(index)
        if not os.path.exists(savepath):
            os.mkdir(savepath)
        cv2.imwrite(filename, img)

###########################################################################################
Code for converting images back to npy files:
from pathlib import Path
from PIL import Image
import numpy as np

root_path = Path("/data/ConvLSTM/movingmnist/")
output_file_path = "/data/ConvLSTM/output.npy"

video_data_list = []

for j in range(10000):  # 假設共有 10000 個影片資料夾
    video_path = root_path / f"video_{j}"
    if video_path.exists():
        image_data_list = []
        image_path_list = list(video_path.glob("*.jpg"))
        for image_file in image_path_list:
            with Image.open(image_file) as img:
                image_data_list.append(img.copy())  
        video_data_list.append(image_data_list)

video_data = np.array(video_data_list)
video_data = video_data.transpose(1, 0, 2, 3)
print("NumPy array 的形狀:", video_data.shape)

np.save(output_file_path, video_data)


The prediction is not one-to-one, but rather a prediction from the previous time step.
![output驗證](https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/assets/151805470/cae744e7-5a9b-4715-b395-ff238d5a8fcd)

**Comment by tsugumi-sys** [comment]
> 1. Since my data is a large npy file merged from npy files of different dates and times, I would like to know what each evaluation.png represents in terms of date and time.

You need to change the evaluation process so that you can link the evaluation results and date.
For example, a simple way is that you load each validation datafiles and save prediction results one by one :)


> 2. If I want to colorize the images and label each frame with the date and time in the file name, do I need to make modifications in the visualize_utils.py section?

Yes you do.
You may need to change `cmap`?
https://github.com/tsugumi-sys/SA-ConvLSTM-Pytorch/blob/1fb3b34874b85b116999ef9c1aac8fc4c2a80047/pipelines/utils/visualize_utils.py#L72


> 3. Regarding the parameters, I would like to ask why "num_kernels" is set to 64?

There is no reason for this, it's just an example. (I remember that the original paper of ConvLSTM uses 64 kernels.)

**Comment by houjanet** [comment]
Thank you for your response, I'll give it a try.

**Comment by houjanet** [comment]
I'm still wondering

Why did you use BCELoss to calculate the loss and use L1Loss to calculate the accuracy?
Have you made any transformations to the moving_mnist dataset so that the model eventually performs classification?

**Comment by tsugumi-sys** [comment]
@houjanet 
BCELoss can be applied not only classification task but also for regression task.
If you apply BCELoss for regression, you need to scale  data so that the minimum value is 0 and the maximum value is 1.


See also https://stats.stackexchange.com/questions/370179/why-binary-crossentropy-can-be-used-as-the-loss-function-in-autoencoders

**Comment by tsugumi-sys** [comment]
I applied sa-convlstm for rainfall prediction (regression) and I normalized data with [0,1] scaling, BCELoss had better performance than other loss functions usually applied for regression tasks.

**Comment by houjanet** [comment]
I see. Thank you for your response.

**Comment by tsugumi-sys** [comment]
@houjanet 
I close this issue :)
If you have any further question, please open a new issue.
